Projects/UOContent/Misc/CharacterCreation.cs

```diff
@@ -256,11 +258,22 @@ private static CityInfo GetStartLocation(CharacterCreatedEventArgs args, bool is
                 }
             case "paladin":
                 {
-                    return _newHavenInfo;
+                    if (Core.AvailableMaps.Includes(MapSelectionFlags.Trammel) && post6000Supported)
+                    {
+                        return _newHavenInfo;
+                    }
+
+                    break;
                 }
             case "samurai":
                 {
-                    if ((flags & ClientFlags.Tokuno) != 0)
+                    bool haotisAndTokunoAccessible = 
+                        ((flags & ClientFlags.Tokuno) == ClientFlags.Tokuno) &&
+                        ((flags & ClientFlags.Malas) == ClientFlags.Malas) &&
+                        Core.AvailableMaps.Includes(MapSelectionFlags.Malas) &&
+                        Core.AvailableMaps.Includes(MapSelectionFlags.Tokuno);
```

Comment:
These can be combined:

```suggestion
                        Core.AvailableMaps.Includes(MapSelectionFlags.Malas | MapSelectionFlags.Tokuno)
```